File: src/sys/java/fan/sys/Facets.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   4 Jul 06  Brian Frank  Creation
//   4 Sep 07  Brian Frank  Rework for new design
//
package fan.sys;

import java.util.Iterator;
import java.util.HashMap;
import java.util.Map.Entry;
import fanx.fcode.*;
import fanx.serial.*;

/**
 * Facets manages facet meta-data as a Str:Obj map.
 */
public final class Facets
{
  public static Facets empty()
  {
    Facets x = emptyVal;
    if (x == null) x = emptyVal = new Facets(new HashMap());
    return x;
  }

  public static Facets makeTransient()
  {
    Facets x = transientVal;
    if (x == null)
    {
      HashMap m = new HashMap();
      m.put(Sys.TransientType, "");
      x = transientVal = new Facets(m);
    }
    return x;
  }

  static Facets mapFacets(Pod pod, FAttrs.FFacet[] ffacets)
  {
    if (ffacets == null || ffacets.length == 0) return empty();
    HashMap map = new HashMap(ffacets.length*3);
    for (int i=0; i<ffacets.length; ++i)
    {
      FAttrs.FFacet ff = ffacets[i];
      Type t = pod.type(ff.type);
      if (!t.isJava()) map.put(t, ff.val);
    }
    return new Facets(map);
  }

  private Facets(HashMap map) { this.map = map; }

  final synchronized List list()
  {
    if (list == null)
    {
      list = new List(Sys.FacetType, map.size());
      Iterator it = map.keySet().iterator();
      while (it.hasNext())
      {
        Type type = (Type)it.next();
        list.add(get(type, true));
      }
      list = (List)list.toImmutable();
    }
    return list;
  }

  final synchronized Facet get(Type type, boolean checked)
  {
    Object val = map.get(type);
    if (val instanceof Facet) return (Facet)val;
    if (val instanceof String)
    {
      Facet f = decode(type, (String)val);
      map.put(type, f);
      return f;
    }
    if (checked) throw UnknownFacetErr.make(type.qname()).val;
    return null;
  }

  final Facet decode(Type type, String s)
  {
    try
    {
      // if no string use make/defVal
      if (s.length() == 0) return (Facet)type.make();

      // decode using normal Fantom serialization
      return (Facet)ObjDecoder.decode(s);
    }
    catch (Throwable e)
    {
      String msg = "ERROR: Cannot decode facet " + type + ": " + s;
      System.out.println(msg);
      e.printStackTrace();
      map.remove(type);
      throw IOErr.make(msg).val;
    }
  }

  final Facets dup()
  {
    return new Facets((HashMap)map.clone());
  }

  final void inherit(Facets facets)
  {
    if (facets.map.size() == 0) return;
    list = null;
    Iterator it = facets.map.entrySet().iterator();
    while (it.hasNext())
    {
      Entry entry = (Entry)it.next();
      Type key = (Type)entry.getKey();

      // if already mapped skipped
      if (map.get(key) != null) continue;

      // if not an inherited facet skip it
      FacetMeta meta = (FacetMeta)key.facet(Sys.FacetMetaType, false);
      if (meta == null || !meta.inherited) continue;

      // inherit
      map.put(key, entry.getValue());
    }
  }

  private static Facets emptyVal;
  private static Facets transientVal;

  private HashMap map;   // Type : String/Facet, lazy decoding
  private List list;     // Facet[]

}

File: src/sys/dotnet/fan/sys/Facets.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   27 Sep 06  Andy Frank  Creation
//   4 Sep 07   Andy Frank  Rework for new design
//

using System.Collections;
using System.Runtime.CompilerServices;
using Fanx.Fcode;
using Fanx.Serial;

namespace Fan.Sys
{
  /// <summary>
  /// Facets manages facet meta-data as a string:Obj map.
  /// </summary>
  public sealed class Facets
  {

    public static Facets empty()
    {
      Facets e = m_empty;
      if (e != null) return e;
      return m_empty = new Facets(new Hashtable());
    }

    public static Facets mapFacets(Pod pod, FAttrs.FFacet[] ffacets)
    {
      if (ffacets == null || ffacets.Length == 0) return empty();
      Hashtable map = new Hashtable();
      for (int i=0; i<ffacets.Length; ++i)
      {
        FAttrs.FFacet ff = ffacets[i];
        Type t = pod.findType(ff.type);
        map[t] = ff.val;
      }
      return new Facets(map);
    }

    private Facets(Hashtable map) { this.m_map = map; }

    [MethodImpl(MethodImplOptions.Synchronized)]
    public List list()
    {
      if (m_list == null)
      {
        m_list = new List(Sys.FacetType, m_map.Count);
        IDictionaryEnumerator en = ((Hashtable)m_map.Clone()).GetEnumerator();
        while (en.MoveNext())
        {
          Type type = (Type)en.Key;
          m_list.add(get(type, true));
        }
        m_list = (List)m_list.toImmutable();
      }
      return m_list;
    }

    [MethodImpl(MethodImplOptions.Synchronized)]
    public Facet get(Type type, bool check)
    {
      object val = m_map[type];
      if (val is Facet) return (Facet)val;
      if (val is string)
      {
        Facet f = decode(type, (string)val);
        m_map[type] = f;
        return f;
      }
      if (check) throw UnknownFacetErr.make(type.qname()).val;
      return null;
    }

    [MethodImpl(MethodImplOptions.Synchronized)]
    public Facet decode(Type type, string s)
    {
      try
      {
        // if no string use make/defVal
        if (s.Length == 0) return (Facet)type.make();

        // decode using normal Fantom serialization
        return (Facet)ObjDecoder.decode(s);
      }
      catch (System.Exception e)
      {
        string msg = "ERROR: Cannot decode facet " + type + ": " + s;
        System.Console.WriteLine(msg);
        Err.dumpStack(e);
        m_map.Remove(type);
        throw IOErr.make(msg).val;
      }
    }


    private static Facets m_empty;

    private Hashtable m_map;   // Type : String/Facet, lazy decoding
    private List m_list;       // Facet[]

  }

/*

  //////////////////////////////////////////////////////////////////////////
  // FCode
  //////////////////////////////////////////////////////////////////////////

    public static MapType mapType()
    {
      MapType t = m_mapType;
      if (t != null) return t;
//      return m_mapType = new MapType(Sys.SymbolType, Sys.ObjType.toNullable());
return null;
    }

    public static Facets empty()
    {
      Facets e = m_empty;
      if (e != null) return e;
      return m_empty = new Facets(new Hashtable());
    }

    /// <summary>
    /// This is the constructor used during decoding the pod
    /// file. The values are all passed in as encoded Strings.
    /// </summary>
    public static Facets make(Hashtable src)
    {
      if (src == null || src.Count == 0) return empty();
      return new Facets(src);

    }

  //////////////////////////////////////////////////////////////////////////
  // Private Constructor
  //////////////////////////////////////////////////////////////////////////

    private Facets(Hashtable map) { this.m_map = map; }

  //////////////////////////////////////////////////////////////////////////
  // Access
  //////////////////////////////////////////////////////////////////////////


    [MethodImpl(MethodImplOptions.Synchronized)]
    public Facet get(Type qname, bool check)
    {
      object val = m_map[qname];
      if (val == null) return def;

      // if we've already decoded, go with it
      if (!(val is Symbol.EncodedVal)) return val;

      // decode into an object
      object obj = Symbol.decodeVal((Symbol.EncodedVal)val);

      // if the object is immutable, then it safe to cache
      if (FanObj.isImmutable(obj)) m_map[qname] = obj;

      return obj;
    }

    [MethodImpl(MethodImplOptions.Synchronized)]
    public List list()
    {
      return null;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    private static MapType m_mapType;
    private static Facets m_empty;
    private static Map m_emptyMap;

    private Hashtable m_map;
  }
  */
}

